### ISSUE TITLE
Make the read connection anonymous.

### ISSUE THREAD

**Issue by Mm2PL** [Chatterino/chatterino2#1817]
**What should be added?**
I would like to have the read connection made always anonymous, not the current account. This could solve an issue that is caused by how account switching works.

**Pros and cons of this solution**
  1. This allows to not reconnect the read connection when changing accounts.
  2. Whisper handling would have to be moved to the write connection.
  3. The biggest problem: Would break Twitch's "you can't join if you're banned" feature.

**Comment by RAnders00** [comment]
This would also mean you don't show up in channel's chatters lists which is not desireable

**Comment by RAnders00** [comment]
Additionally, when a channel is joined, the server communicates user-specific settings for that channel, for example whether the user is a moderator there (which Chatterino uses to show the mod buttons). Changing it to anonymous would mean we would have to remove the mod button or show it all the time, or implement a complicated mechanism that calls `/mods` or similar.

**Comment by pajlada** [comment]
Yeah I don't think there are any real pros that outweigh the cons in this switch, won't fix

**Comment by zneix** [comment]
Looked recently at the code responsible for handling IRC messages and:

> This would also mean you don't show up in channel's chatters lists which is not desireable

We already send `JOIN`s and `PART`s on both write and read connections, so I think nothing would change if we were to `JOIN` channels only on write connection and correct me if I'm wrong, but wouldn't we still show up in chatters list?
Proof that we `JOIN` the same channel twice:
![fdm](https://cdn.zneix.eu/mTcSZQs.png)

> Additionally, when a channel is joined, the server communicates user-specific settings for that channel (USERSTATE message),

Played a lot with `USERSTATE` recently and from what I know you can receive that message after either `JOIN`ing a channel or sending a message.
We would still get `USERSTATE` after `JOIN`ing channels on the write connection and in addition to that, we can still get `USERSTATE` after we send a message (fun fact: we receive it even if we're not `JOIN`ed to the channel).

So given that, I think there's actually no cons, or rather no cons that we can't solve but pros as described by mm2pl in issue body are worth performing the switch imho. The logic of how do we handle certain IRC messages would just need to be adjusted.
@pajlada let me know what you think and if what I said makes sense please reopen and I can assign myself and try to make this work.

**Comment by pajlada** [comment]
>The biggest problem: Would break Twitch's "you can't join if you're banned" feature.

This is still true, so I still don't think this is a functionality we should look to add.

**Comment by apa420** [comment]
I don't really understand the reasoning for not being able to join the chat when banned and I don't think chatterino should follow exactly how twitch chat does it, because that's not what we do always anyways.

If this feature is really really really really really important we can on the write connection see that we can't join and are banned and just stop showing messages in that split.

The point of doing the read connection anonymous is to make people be able to switch accounts without having to reconnect to chat and miss messages.

**Comment by zneix** [comment]
@pajlada please have a look at #3107. That rate-limit I described in the issue has been in docs for quite some time already. It was "cosmetic" until twitch suddenly fixed that today, so we had no problems until yesterday.

For that issue, throttling would have to be implemented. Thing is - with having to JOIN many channels on both connections, it'd be ridiculously slow as waiting to fully join all the channels on both connections.

Example: I have 62 unique chats added to my Chatterino. It means that I'd need to send 124 JOINs. With a future fix it'll need to be split into 7 batches of JOIN messages with 10 seconds of cooldown before executing next batch, resulting in 60 seconds of wait before I can JOIN all the chats on both connections.
If we'd make only 1 connection JOIN channels on the user's account, it'd reduce that time to 30 seconds. While it's still a lot, it's way better in my opinion. It'd also grant us all other benefits described by @mm2pl in the issue an my recent comment.
While I fully understand and respect your opinion/concerns about the point 3. making Chatterino the "ban evasion app", I think that JOINing as little chats as possible is more important than that.


I'll reopen this for the time being to keep the conversation going and make issue more noticable as it's relevant given the current situation.

**Comment by LosFarmosCTL** [comment]
I am HIGHLY in favor of this, solves basically all issues I had with the recent JOIN rate limit changes.

This would not only make starting chatterino and especially switching accounts finally not completely suck anymore, but also make issues like #1308 a lot less relevant, since no messages would be lost on switching accounts etc., instead only if the read connection actually terminates, which, at least in my experience, is a very infrequent issue.

And, reiterating what others have said, since chatterino has an anonymous connection mode anyways, I really dont see a huge issue in banned channels being readable while the user is connected with their real account, after all banning someone from your chat is basically just about stopping them from sending messages. If someone wants to read messages in a chat he is banned in, it really doesn't take chatterino to do that, even just opening twitch in a private browser window would be enough.

**Comment by Nerixyz** [comment]
I took a shot at a prototype in https://github.com/Nerixyz/chatterino2/commit/c12f0e9e5e781eb8cdbe926c8eab2c43e1a3f708, and I hope to land that incrementally. There it's optional where you can choose to use multiple connections (& anonymous mode). This significantly speeds up the startup process. Afterward, the bottleneck is loading recent messages (we could do that in parallel).

**Comment by pajlada** [comment]
This still bypasses channel bans - after 5 years this is still not something I think is worth implementing for that reason alone.

**Comment by iProdigy** [comment]
> This still bypasses channel bans - after 5 years this is still not something I think is worth implementing for that reason alone.

ugly idea: do fast joins with anon read connection & use another mechanism to unjoin the channels where banned (e.g., authed read connection or repeatedly creating/deleting eventsub chat ws subscriptions... hopefuly we can think of something more elegant because these 2 ideas are probably not worth the complexity)

**Comment by Nerixyz** [comment]
> This still bypasses channel bans - after 5 years this is still not something I think is worth implementing for that reason alone.

Both https://github.com/Chatterino/chatterino2/issues/1817#issuecomment-886182291 and https://github.com/Chatterino/chatterino2/issues/1817#issuecomment-908496344 make the point that Chatterino already has a mode to join chat anonymously.

---

> repeatedly creating/deleting eventsub chat ws subscriptions

I didn't think of that, but I think it's the "best" solution we have, since it gives us an easy way of figuring out if the subscription failed (an HTTP error).

> probably not worth the complexity

I agree. Though, if anyone else wants to take a shot at this, feel free to use my commit as a base.

**Comment by pajlada** [comment]
I disagree with both earlier comments - changing your account in Chatterino is the same vane as opening the Twitch channel in a different browser or incognito mode.

Us allowing you to be read the chat & interact as-if you weren't banned, with no penatly at all, makes this feature bypass channel ban.
For this to be implemented, it would need some feature like iProdigy mentioned, where we keep track of channels you're banned in and stop showing messages there, which does not seem worth the added complexity.

**Comment by pajlada** [comment]
I am closing this as wont-fix again. Feel free to continue discussion as needed, but this is not something I'm looking to have implemented at the current time. Bypassing Twitch's moderation rules (or the added complexity needed to re-institute them) is not worth it to me.
